src/Nethermind/Nethermind.Consensus/Producers/TxPoolTxSource.cs

```diff
@@ -77,52 +122,81 @@ public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimi
+                foreach (Transaction blobTx in selectedBlobTxs)
+                {
+                    yield return blobTx;
+                }
+            }
+
+            if (_logger.IsDebug) _logger.Debug($"Potentially selected {selectedTransactions} out of {i} pending transactions checked.");
+        }
+
+        private bool TryGetFullBlobTx(Transaction blobTx, [NotNullWhen(true)] out Transaction? fullBlobTx)
+        {
+            if (blobTx.NetworkWrapper is not null)
+            {
+                fullBlobTx = blobTx;
+                return true;
+            }
+            fullBlobTx = null;
+            return blobTx.Hash is not null && _transactionPool.TryGetPendingBlobTransaction(blobTx.Hash, out fullBlobTx);
+        }
+
+        private bool TryUpdateBlobGasPrice(Transaction fullBlobTx, BlockHeader parent, out UInt256 blobGasPrice)
+        {
+            ulong? excessDataGas = BlobGasCalculator.CalculateExcessBlobGas(parent, _specProvider.GetSpec(parent));
+            if (excessDataGas is null)
+            {
+                if (_logger.IsTrace) _logger.Trace($"Declining {fullBlobTx.ToShortString()}, the specification is not configured to handle shard blob transactions.");
+                blobGasPrice = UInt256.Zero;
+                return false;
+            }
+            if (!BlobGasCalculator.TryCalculateBlobGasPricePerUnit(excessDataGas.Value, out blobGasPrice))
+            {
+                if (_logger.IsTrace) _logger.Trace($"Declining {fullBlobTx.ToShortString()}, failed to calculate data gas price.");
+                blobGasPrice = UInt256.Zero;
+                return false;
+            }
+            return true;
+        }
 
-                    int txAmountOfBlobs = tx.BlobVersionedHashes?.Length ?? 0;
+        private IEnumerable<Transaction> PickBlobTxsBetterThanCurrentTx(ArrayPoolList<Transaction> selectedBlobTxs, Transaction tx, IComparer<Transaction> comparer)
+        {
+            if (selectedBlobTxs.Count > 0)
+            {
+                using ArrayPoolList<Transaction> txsToRemove = new(selectedBlobTxs.Count);
 
-                    if (blobGasPrice > tx.MaxFeePerBlobGas)
+                foreach (Transaction blobTx in selectedBlobTxs)
+                {
+                    if (comparer.Compare(blobTx, tx) > 0)
                     {
-                        if (_logger.IsTrace) _logger.Trace($"Declining {tx.ToShortString()}, blob gas fee is too low.");
-                        continue;
+                        yield return blobTx;
+                        txsToRemove.Add(blobTx);
                     }
-
-                    if (BlobGasCalculator.CalculateBlobGas(blobsCounter + txAmountOfBlobs) >
-                        Eip4844Constants.MaxBlobGasPerBlock)
+                    else
                     {
-                        if (_logger.IsTrace) _logger.Trace($"Declining {tx.ToShortString()}, no more blob space.");
-                        continue;
+                        break;
```

Comment:
I don't get it. First selected blob tx is the best, so before yield returning non-blob tx we are checking if there is a better blob tx to yield return first. If it is, we are yield returning it, removing from collection of selected blob txs and checking again with another blob tx - best remaining one. If best blob tx is not better, then we are not comparing worse ones - so there is only one unsuccessful check or successful yield return and further check/checks. If we will do it in backward order as proposed in `a)`, we would need to iterate collection in case of `tx > worst_blob` until we will find `tx < blob_tx` and then yield return in initial order. Seems like unnecessary complicating